# Initialized

In [1]:
# At end of session
import duckdb as ddb
ddb.close()

In [2]:
import os
from dotenv import load_dotenv
import pandas as pd
import duckdb as ddb
import requests
import datetime
import time
import pandas_ta as ta

from SQL.Querys import DBconn
con = ddb.connect(r'C:\Design Folder\RBGithub\Trading_Algo_Data\testingalphaoptions.duckdb')
db = DBconn(con)

date = '2024-07-05'

### Get the Stock Data form Alpha Vantage

In [3]:
load_dotenv()
api_key = os.getenv("Alpha_API")
symbol = 'AAPL'
interval = '1'
Stock_lookback_period = 100
Increasing_Mins = 2

#url = f"https://www.alphavantage.co/query?function=TIME_SERIES_INTRADAY&symbol={symbol}&interval=1min&month=2024-07&outputsize=full&apikey={api_key}"

stock_api = f'https://www.alphavantage.co/query?function=TIME_SERIES_INTRADAY&symbol={symbol}&interval={interval}min&entitlement=realtime&outputsize=full&apikey={api_key}'


r = requests.get(stock_api)
key = f'Time Series (1min)'
data = r.json().get(key, {})

df = pd.DataFrame.from_dict(data, orient='index')

df.index = pd.to_datetime(df.index)

stock_df = df.loc[date]

new_column_names = ['open', 'high', 'low', 'close', 'volume']
stock_df.columns = new_column_names


stock_df = stock_df.copy()
stock_df['timestamp'] = stock_df.index
stock_df.reset_index(drop=True,inplace=True)

stock_df['Trade'] = False
stock_df['close'] = pd.to_numeric(stock_df['close'])
stock_df['open'] = pd.to_numeric(stock_df['open'])
clean_stock_df = stock_df



print(stock_df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 960 entries, 0 to 959
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   open       960 non-null    float64       
 1   high       960 non-null    object        
 2   low        960 non-null    object        
 3   close      960 non-null    float64       
 4   volume     960 non-null    object        
 5   timestamp  960 non-null    datetime64[ns]
 6   Trade      960 non-null    bool          
dtypes: bool(1), datetime64[ns](1), float64(2), object(3)
memory usage: 46.1+ KB
None


# Analyze Stock Data 

## EMA and Local Mins Strategy 
1. Stock_df index Starts with newest stock price
2. Swoosh
   - Price should be greater than 4
   - 1 should be greater than 3
   - 3 should be greater than 2

### Notes
1. Swoosh Points tell us When to look at the data 
2. 94 / 959 are Swoosh Points
3. Analysing Local Minmums
   1. Find all the minimums
   2. Check to see if the last X minimums are increases
   3. Check if the last X minimums are > 8 EMA
   4. Check if for the last X Minimums the 8 EMA > 20 EMA

In [20]:
stock_df['Min_Test'] = False
stock_df['EmaMin_Test'] = False
stock_df['EMA_Test'] = False
stock_df['Trade'] = False


stock_df['less1'] = stock_df['close'].shift(-1)
stock_df['less2'] = stock_df['close'].shift(-2)
stock_df['less3'] = stock_df['close'].shift(-3)
stock_df['Swoosh'] = (stock_df['less1']<stock_df['close']) & (stock_df['less2']<stock_df['less1']) & (stock_df['less2']<stock_df['less3'])

swoosh_points = stock_df[stock_df['Swoosh']]

for index, row in swoosh_points.iterrows():
    if index <= index+Stock_lookback_period: 

        table_range = stock_df.iloc[index:index+Stock_lookback_period-1].copy()

        # Stock Analysis 
        # Reverse table to calculate EMA correctly
        table_range_reversed = table_range.iloc[::-1].copy()
        table_range_reversed['EMA8'] = ta.ema(close=table_range_reversed['close'], length=8)
        table_range_reversed['EMA20'] = ta.ema(close=table_range_reversed['close'], length=20)

        # Reverse the EMA columns back to original order
        table_range['EMA8'] = table_range_reversed['EMA8'].iloc[::-1].values
        table_range['EMA20'] = table_range_reversed['EMA20'].iloc[::-1].values

        table_range['Local_Min'] = table_range['close'][(table_range['close'].shift(-1)>table_range['close'])&
                                                        (table_range['close'].shift(1)>table_range['close'])]

        local_mins = table_range[['close','Local_Min','EMA8','EMA20']].dropna()

        if len(local_mins) < 2:
            continue

        local_mins['Min_Test'] = local_mins['Local_Min'] > local_mins['Local_Min'].shift(-1)
        local_mins['EmaMin_Test'] = local_mins['Local_Min'] > local_mins['EMA8']
        local_mins['EMA_Test'] = local_mins['EMA8'] > local_mins['EMA20']
        
        Update_df = local_mins[['Min_Test','EmaMin_Test','EMA_Test']]

        stock_df.update(Update_df)
        table_range.update(Update_df)

        if all([local_mins['Min_Test'].iloc[:1].all(),
                local_mins['EmaMin_Test'].iloc[:1].all(),
                local_mins['EMA_Test'].iloc[:1].all()]):
            stock_df.at[index, 'Trade'] = True
            table_range.at[index, 'Trade'] = True


with pd.ExcelWriter('stockdata3.xlsx') as writer:
    stock_df.to_excel(writer, sheet_name='Sheet1', index=True)

        


print(stock_df[stock_df['Trade']].info())





<class 'pandas.core.frame.DataFrame'>
Index: 25 entries, 43 to 750
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   open              25 non-null     float64       
 1   high              25 non-null     object        
 2   low               25 non-null     object        
 3   close             25 non-null     float64       
 4   volume            25 non-null     object        
 5   timestamp         25 non-null     datetime64[ns]
 6   Trade             25 non-null     bool          
 7   WasRed            25 non-null     bool          
 8   Engulfing Candle  25 non-null     bool          
 9   EMA100            25 non-null     float64       
 10  Found_EMA100      25 non-null     int64         
 11  Min_Test          25 non-null     object        
 12  EmaMin_Test       25 non-null     object        
 13  EMA_Test          25 non-null     object        
 14  less1             25 non-null  

## RSI and Engulfing Candle (REC) Strategy 

In [9]:
REC_df = clean_stock_df

Max_index = len(REC_df['timestamp'])-100

# Calculate Engulfing Candles
REC_df['WasRed'] = REC_df['close'].shift(-1) < REC_df['open'].shift(-1)
REC_df['Engulfing Candle'] = REC_df['WasRed'] & (REC_df['open'] < REC_df['close'].shift(-1)) & (REC_df['close'] > REC_df['open'].shift(-1))

REC_df_Reversed = REC_df.iloc[::-1].copy()
REC_df_Reversed['EMA100'] = ta.ema(close=REC_df_Reversed['close'], length=100)
REC_df['EMA100'] = REC_df_Reversed['EMA100'].iloc[::-1].values

REC_df['Found_EMA100'] = 0
REC_df['RSI'] = 0

Engulfing_df = REC_df[REC_df['Engulfing Candle']]


for index, row in Engulfing_df.iterrows():
    if index < Max_index:
        Engulfing_range = REC_df.iloc[index:index+100].copy()

        Engulfing_range_reversed = Engulfing_range.iloc[::-1].copy()
        Engulfing_range_reversed['EMA100'] = ta.ema(close=Engulfing_range_reversed['close'], length=100)
        Engulfing_range_reversed['RSI'] = ta.rsi(close=Engulfing_range_reversed['close'])

        Engulfing_range['EMA100'] = Engulfing_range_reversed['EMA100'].iloc[::-1].values
        Engulfing_range['RSI'] = Engulfing_range_reversed['RSI'].iloc[::-1].values

        REC_df['Found_EMA100'].update(Engulfing_range['EMA100'])
        REC_df['RSI'].update(Engulfing_range['RSI'])


        if Engulfing_range['close'].iloc[0] > Engulfing_range['EMA100'].iloc[0] + (Engulfing_range['EMA100'].iloc[0] * .05):
            if Engulfing_range['RSI'] > 55:
                REC_df.at[index, 'Trade'] = True

Tradable_df = REC_df[REC_df['Trade']]

print(Tradable_df.info())
        
with pd.ExcelWriter('RECdata.xlsx') as writer:
    REC_df.to_excel(writer, sheet_name='Sheet1', index=True)


<class 'pandas.core.frame.DataFrame'>
Index: 23 entries, 44 to 713
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   open              23 non-null     float64       
 1   high              23 non-null     object        
 2   low               23 non-null     object        
 3   close             23 non-null     float64       
 4   volume            23 non-null     object        
 5   timestamp         23 non-null     datetime64[ns]
 6   Trade             23 non-null     bool          
 7   WasRed            23 non-null     bool          
 8   Engulfing Candle  23 non-null     bool          
 9   EMA100            23 non-null     float64       
 10  Found_EMA100      23 non-null     float64       
 11  RSI               23 non-null     float64       
dtypes: bool(3), datetime64[ns](1), float64(5), object(3)
memory usage: 1.9+ KB
None


C:\Users\rybot\AppData\Local\Temp\ipykernel_27952\1578477649.py:30: FutureWarning:

Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[226.3518]' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.

C:\Users\rybot\AppData\Local\Temp\ipykernel_27952\1578477649.py:31: FutureWarning:

Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[62.80110014 58.8617063  61.58701887 55.40978832 39.94939648 44.22051181
 45.73394698 48.8381077  55.8821867  47.01112944 48.83475526 44.8625958
 46.54010254 48.21415776 44.50713821 44.50713821 47.43494826 50.5209794
 53.76924179 43.68226298 46.84597103 44.99641457 53.66641952 53.66641952
 53.66641952 57.19591291 51.24046205 51.24046205 54.51015892 54.51015892
 54.51015892 54.51015892 51.02501003 51.02501003 52.17387609 50.08675615
 50.08675615 56.46023901 50.62625396 52.81676334 52.81676334 46.87030262
 48.77288438 47.7

### Getting Data From Options DB 
1. Get a List of Options Log Times
2. Write a For Loop that goes through 

In [ ]:
test_df = stock_df.copy()

table_range_reversed = test_df.iloc[::-1].copy()
table_range_reversed['EMA8'] = ta.ema(close=table_range_reversed['close'], length=8)
table_range_reversed['EMA20'] = ta.ema(close=table_range_reversed['close'], length=20)

# Reverse the EMA columns back to original order
test_df['EMA8'] = table_range_reversed['EMA8'].iloc[::-1].values
test_df['EMA20'] = table_range_reversed['EMA20'].iloc[::-1].values

test_df['Local_Min'] = test_df['close'][(test_df['close'].shift(-1)>test_df['close'])&
                                                (test_df['close'].shift(1)>test_df['close'])]


with pd.ExcelWriter('Stockdf2.xlsx') as writer:
  test_df.to_excel(writer, sheet_name='Sheet1', index=True)